### PR TITLE
[DEE-510] nlp integrity checks

### PR DESCRIPTION
**Pull request by yromanyshyn**

### CODE REVIEW

File: deepchecks/nlp/checks/data_integrity/text_duplicates.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        self.remove_stopwords = remove_stopwords
+        self.n_to_show = n_to_show
+        self.n_samples = n_samples
+        self.random_state = random_state
+
+    def run_logic(self, context: Context, dataset_kind):
+        """Run check."""
+        dataset = context.get_data_by_kind(dataset_kind).sample(self.n_samples, random_state=self.random_state)
+        dataset = t.cast(TextData, dataset)
+        samples = dataset.text
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/data_integrity/text_duplicates.py:66]: redundant variable - only used once

File: deepchecks/nlp/checks/data_integrity/text_duplicates.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        }
+
+        if not (context.with_display and percent_of_duplicates > 0):
+            return CheckResult(value=result_value)
+
+        first_sample = grouped_samples['Text'].first()
+        instances = grouped_samples['Sample ID'].aggregate(lambda x: format_list(x.to_list()))
+
+        # TODO: refactor
+        table = pd.DataFrame({
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/data_integrity/text_duplicates.py:112]: For all checks add parameter - max_text_length_for_display
and in displays if the text is longer than cut it after max_text_length_for_display and add ...

File: deepchecks/nlp/checks/data_integrity/text_duplicates.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+               'Each row in the table shows an example of a text duplicate and the number of times it appears.',
+               table
+            ]
+        )
+
+
+T = t.TypeVar("T")
+
+
+def to_ordional_enumeration(data: t.List[T]) -> t.Dict[T, int]:
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/data_integrity/text_duplicates.py:133]: make private + add some docstring what it does

File: deepchecks/utils/abstracts/data_duplicates.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+from typing_extensions import Self
+
+from deepchecks.core import ConditionCategory, ConditionResult
+from deepchecks.utils.strings import format_percent
+
+__all__ = ['DataDuplicatesAbstract']
+
+
+class DataDuplicatesAbstract(abc.ABC):
```
Comment (Nadav-Barak) [inline deepchecks/utils/abstracts/data_duplicates.py:23]: Optional - See if you can add more stuff in the abstract class:
I think the nlp and tabular check display + value can be the same after you get the list of duplicated indexes

File: deepchecks/nlp/checks/train_test_validation/train_test_sample_mix.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+from deepchecks.core import CheckResult
+from deepchecks.core.errors import DeepchecksValueError
+from deepchecks.nlp import Context, TrainTestCheck
+from deepchecks.utils.strings import format_percent
+from deepchecks.nlp.utils.text_utils import normalize_text, hash_text
+from deepchecks.utils.strings import format_list, format_percent
+from deepchecks.utils.abstracts.train_test_samples_mix import TrainTestSamplesMixAbstract
+from deepchecks.nlp.text_data import TextData
+from deepchecks.nlp.checks.data_integrity.text_duplicates import to_ordional_enumeration
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/train_test_validation/train_test_sample_mix.py:23]: if after refactor still used in both checks move it to utils

File: deepchecks/nlp/checks/train_test_validation/train_test_sample_mix.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        self.ignore_case = ignore_case
+        self.remove_punctuation = remove_punctuation
+        self.normalize_unicode = normalize_unicode
+        self.remove_stopwords = remove_stopwords
+        self.ignore_whitespace = ignore_whitespace
+        self.n_samples = n_samples
+        self.n_to_show = n_to_show
+        self.random_state = random_state
+
+    def _prepare_hashes(self, text: t.Iterable[str]) -> t.List[int]:
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/train_test_validation/train_test_sample_mix.py:64]: in utils + use also in data duplicates

File: deepchecks/utils/abstracts/train_test_samples_mix.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+from typing_extensions import Self
+
+from deepchecks.core import ConditionCategory, ConditionResult
+from deepchecks.utils.strings import format_percent
+
+__all__ = ['TrainTestSamplesMixAbstract']
+
+
+class TrainTestSamplesMixAbstract(abc.ABC):
```
Comment (Nadav-Barak) [inline deepchecks/utils/abstracts/train_test_samples_mix.py:23]: Also here see if you can add more common logic - specifically the display and value calculation based on some common format

File: deepchecks/nlp/checks/data_integrity/text_duplicates.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+        value['Duplicate'] = value['Duplicate'].apply(lambda x: duplicates_enumeration[x])
+        value = value.set_index(["Duplicate", "Sample ID"])
+
+        result_value = {
+            "percent_of_duplicates": percent_of_duplicates,
+            "duplicates": value
+        }
+
+        if not (context.with_display and percent_of_duplicates > 0):
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/nlp/checks/data_integrity/text_duplicates.py:105]: what exactly?
this line?

```
if not (context.with_display and percent_of_duplicates > 0):
```

why?

File: deepchecks/nlp/checks/data_integrity/text_duplicates.py (hunk 1 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/data_integrity/text_duplicates.py:83]: currently, it is used twice

File: deepchecks/nlp/checks/data_integrity/text_duplicates.py (hunk 2 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/data_integrity/text_duplicates.py:118]: done

File: deepchecks/nlp/checks/data_integrity/text_duplicates.py (hunk 3 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/data_integrity/text_duplicates.py]: moved to utils, and added docstring

File: deepchecks/nlp/checks/train_test_validation/train_test_sample_mix.py (hunk 5 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/train_test_validation/train_test_sample_mix.py]: moved

File: deepchecks/utils/abstracts/data_duplicates.py (hunk 4 above)
Comment (yromanyshyn) [reply deepchecks/utils/abstracts/data_duplicates.py:23]: I suggest creating a separate issue for this and all other checks

The problem is that the checks work similarly but not quite similarly/
It is not obvious how to extract/separate their common logic into a generic function that will be useful, clean, and easy to use in both checks

so, currently, I would prefer how it is

File: deepchecks/utils/abstracts/train_test_samples_mix.py (hunk 7 above)
Comment (yromanyshyn) [reply deepchecks/utils/abstracts/train_test_samples_mix.py:23]: look at the [comment](https://github.com/deepchecks/deepchecks/pull/2471#discussion_r1178681658) above

File: deepchecks/nlp/checks/data_integrity/conflicting_labels.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        percent_of_ambiguous_samples = num_of_ambiguous_samples / n_of_samples
+
+        result_df = ambiguous_samples.rename(columns={'hash': 'Duplicate'})
+        duplicates_enumeration = to_ordional_enumeration(result_df['Duplicate'].to_list())
+        result_df['Duplicate'] = result_df['Duplicate'].apply(lambda x: duplicates_enumeration[x])
+        result_df = result_df.set_index(['Duplicate', 'Sample ID', 'Label'])
+
+        result_value = {
+            'percent': percent_of_ambiguous_samples,
+            'ambiguous_samples': result_df,
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/conflicting_labels.py:131]:
```suggestion
            'conflicting_samples': result_df,
```

File: deepchecks/nlp/checks/data_integrity/conflicting_labels.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        num_of_ambiguous_samples = ambiguous_samples['Text'].count()
+        percent_of_ambiguous_samples = num_of_ambiguous_samples / n_of_samples
+
+        result_df = ambiguous_samples.rename(columns={'hash': 'Duplicate'})
+        duplicates_enumeration = to_ordional_enumeration(result_df['Duplicate'].to_list())
+        result_df['Duplicate'] = result_df['Duplicate'].apply(lambda x: duplicates_enumeration[x])
+        result_df = result_df.set_index(['Duplicate', 'Sample ID', 'Label'])
+
+        result_value = {
+            'percent': percent_of_ambiguous_samples,
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/conflicting_labels.py:130]:
```suggestion
            'percent_conflicting_samples': percent_of_ambiguous_samples,
```

File: deepchecks/nlp/checks/data_integrity/conflicting_labels.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+        by_hash = ambiguous_samples.groupby(['hash'], dropna=False)
+        fn = lambda x: format_list(x.to_list())
+        observed_labels = by_hash['Label'].aggregate(fn)
+        instances = by_hash['Sample ID'].aggregate(fn)
+        first_in_group = by_hash['Text'].first()
+
+        display_table = (
+            pd.DataFrame({
+                # TODO:
+                # for multi-label and token classification
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/conflicting_labels.py:147]: Can we start by displaying the first two non-equal things and then "..."?

File: deepchecks/nlp/checks/data_integrity/conflicting_labels.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+                'Observed Labels': observed_labels,
+                'Instances': instances,
+                'Text': first_in_group
+            })
+            .reset_index(drop=True)
+            .set_index(['Observed Labels', 'Instances'])
+        )
+        table_description = (
+            'Each row in the table shows an example of a data sample '
+            'and the its observed labels as found in the dataset.'
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/conflicting_labels.py:160]:
```suggestion
            'and the its observed conflicting labels as found in the dataset.'
```

File: deepchecks/nlp/checks/data_integrity/text_duplicates.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/checks/data_integrity/text_duplicates.py:114]: why the "not"? We want to display if both conditions ARE true.

File: deepchecks/nlp/checks/data_integrity/text_duplicates.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+            'Number of Samples': counted_duplicates
+        })
+        table = table.iloc[:self.n_to_show]
+        table = table.sort_values(by=['Number of Samples'], ascending=False)
+        table = table.set_index(['Instances', 'Number of Samples'])
+
+        return CheckResult(
+            value=result_value,
+            display=[
+                f'{format_percent(percent_of_duplicates)} of data samples are duplicates.',
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/text_duplicates.py:133]:
```suggestion
                f'{format_percent(percent_of_duplicates)} of data samples are duplicates.\n',
```

File: deepchecks/nlp/checks/train_test_validation/train_test_sample_mix.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+            'Test text sample': first_sample_in_group.apply(self._truncate_text),
+            'Number of test duplicates': counted_test_duplicates
+        })
+
+        display_table = display_table.iloc[:self.n_to_show]
+        display_table = display_table.reset_index(drop=True).set_index(['Train instances', 'Test instances'])
+
+        message = (
+            f'{format_percent(duplicates_ratio)} ({n_of_test_duplicates} / {n_of_test_samples}) '
+            'of test data samples appear in train data'
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/train_test_sample_mix.py:157]:
```suggestion
            'of test data samples also appear in train data'
```

File: deepchecks/utils/abstracts/data_duplicates.py (hunk 4 above)
Comment (noamzbr) [reply deepchecks/utils/abstracts/data_duplicates.py:23]: So Yuri can you open that issue please?

File: deepchecks/nlp/checks/data_integrity/special_characters.py [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+        display_table = (
+            display_table.sort_values(by=['% of Samples With Character'], ascending=False)
+            .reset_index(drop=True)
+            .set_index(['Special Character'])
+        )
+        if self.n_most_common > display_table.shape[0]:
+            message = ''
+        else:
+            message = (
+                f'Showing only the top {self.n_most_common} most common characters, '
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/special_characters.py:143]:
```suggestion
                f'Showing only the top {self.n_most_common} most common special characters, '
```

File: deepchecks/nlp/checks/data_integrity/special_characters.py [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+                'you can change it using n_most_common param'
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/special_characters.py:144]:
```suggestion
                'you can change it using n_most_common param.'
```

File: deepchecks/nlp/checks/data_integrity/special_characters.py [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+            )
+        return CheckResult(
+            value=data,
+            display=[message, display_table.iloc[:self.n_most_common]]
+        )
+
+    # TODO:
+    # is default max_ratio good?
+    # are method and condition names good?
+    def add_condition_ratio_of_special_characters_less_or_equal(self: Self, max_ratio: float = 0.05) -> Self:
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/special_characters.py:154]: I think an additional condition that perhaps is better as a default if on the total % of samples that have any of those special characters.

File: deepchecks/utils/abstracts/data_duplicates.py (hunk 4 above)
Comment (yromanyshyn) [reply deepchecks/utils/abstracts/data_duplicates.py:23]: created
[DEE-520](https://linear.app/deepchecks/issue/DEE-520/common-logic-abstract-for-tabular-and-nlp-integrity-checks)

File: deepchecks/nlp/checks/data_integrity/text_duplicates.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/nlp/checks/data_integrity/text_duplicates.py]: modified

File: deepchecks/nlp/utils/text_utils.py [hunk 18]
```diff
@@ -9,8 +9,25 @@
...
+import typing as t
+import unicodedata
 
-__all__ = ['break_to_lines_and_trim']
+import nltk
+from nltk.corpus import stopwords
+from nltk.tokenize import word_tokenize
+
+# TODO:
+# rename this module just to 'text.py',
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/nlp/utils/text_utils.py]: renamed

File: deepchecks/nlp/utils/text_utils.py [hunk 19]
```diff
@@ -47,3 +64,83 @@ def break_to_lines_and_trim(s, max_lines: int = 10, min_line_length: int = 50, m
         if len(s) > 0:
             lines[-1] = lines[-1] + '...'
     return '<br>'.join(lines)
+
+
+# TODO: check option to save this list of words as a constant
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/nlp/utils/text_utils.py]: hm/ ok, let it be as it is

File: deepchecks/nlp/checks/data_integrity/conflicting_labels.py (hunk 9 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/data_integrity/conflicting_labels.py]: renamed

File: deepchecks/nlp/checks/data_integrity/conflicting_labels.py (hunk 10 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/data_integrity/conflicting_labels.py]: renamed

File: deepchecks/nlp/checks/data_integrity/text_duplicates.py (hunk 13 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/data_integrity/text_duplicates.py:133]: no need, each string item in the `display` sequence will be wrapped with HTML `<p>` tag, 
so, they will be displayed on different lines

File: deepchecks/nlp/checks/data_integrity/special_characters.py [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+        max_text_length_for_display: int = 30,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.special_characters_whitelist = (
+            frozenset(special_characters_whitelist)
+            if special_characters_whitelist
+            else frozenset()
+        )
+        self.special_characters = SPECIAL_CHARACTERS.difference(self.special_characters_whitelist)
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/nlp/checks/data_integrity/special_characters.py]: fixed both points

File: deepchecks/nlp/checks/data_integrity/special_characters.py (hunk 17 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/data_integrity/special_characters.py:172]: added additional condition

File: deepchecks/nlp/checks/data_integrity/special_characters.py [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+
+
+@docstrings
+class SpecialCharacters(SingleDatasetCheck):
+    """Find samples that contain special characters.
+
+    Parameters
+    ----------
+    special_characters_whitelist: Union[str, Sequence[str]] , default ' ' + string.punctuation
+        set of special characters to ignore
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/special_characters.py:47]:
```suggestion
        set of special characters to ignore. Punctuation (string.punctuation) is whitelisted by default.
```

File: deepchecks/nlp/checks/data_integrity/special_characters.py [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+            value=result_value,
+            display=[
+                f'List of ignored special characters: {self.special_characters_whitelist}',
+                message,
+                display_table.iloc[:self.n_most_common]
+            ]
+        )
+
+    # TODO:
+    # is default max_ratio good?
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/special_characters.py:174]: remove todo

**Review by noamzbr — APPROVED** [review]
Add new checks (and their conditions) to the default suites.

File: deepchecks/nlp/checks/data_integrity/special_characters.py (hunk 22 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/data_integrity/special_characters.py]: removed

**Comment by yromanyshyn** [comment]
@noamzbr 

> Add new checks (and their conditions) to the default suites.

added

**Pull request merged**